bindings/go/repository/component/resolver/v1/repository.go

```diff
@@ -0,0 +1,306 @@
+package v1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+	"log/slog"
+	"maps"
+	goruntime "runtime"
+	"slices"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+
+	"ocm.software/open-component-model/bindings/go/blob"
+	"ocm.software/open-component-model/bindings/go/configuration/resolvers/v1/matcher"
+	resolverspec "ocm.software/open-component-model/bindings/go/configuration/resolvers/v1/spec"
+	descriptor "ocm.software/open-component-model/bindings/go/descriptor/runtime"
+	"ocm.software/open-component-model/bindings/go/repository"
+	"ocm.software/open-component-model/bindings/go/runtime"
+)
+
+const Realm = "repository/component/resolver"
+
+type ResolverRepository struct {
+	goRoutineLimit int
+
+	repositoryProvider repository.ComponentVersionRepositoryProvider
+	credentialProvider repository.CredentialProvider
+
+	resolvers []*resolverspec.Resolver
+
+	repositoriesForResolverCacheMu sync.RWMutex
+	repositoriesForResolverCache   []repository.ComponentVersionRepository
+
+	matcherCacheMu sync.RWMutex
+	matcherCache   []*matcher.ResolverMatcher
+}
+
+type ResolverRepositoryOption func(*ResolverRepositoryOptions)
+
+func WithGoRoutineLimit(numGoRoutines int) ResolverRepositoryOption {
+	return func(options *ResolverRepositoryOptions) {
+		options.GoRoutineLimit = numGoRoutines
+	}
+}
+
+type ResolverRepositoryOptions struct {
+	GoRoutineLimit int
+}
+
+func NewResolverRepository(_ context.Context, repositoryProvider repository.ComponentVersionRepositoryProvider, credentialProvider repository.CredentialProvider, res []*resolverspec.Resolver, opts ...ResolverRepositoryOption) (*ResolverRepository, error) {
+	options := &ResolverRepositoryOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if options.GoRoutineLimit <= 0 {
+		options.GoRoutineLimit = goruntime.NumCPU()
+	}
+
+	resolvers := deepCopyResolvers(res)
+
+	matchers := make([]*matcher.ResolverMatcher, len(resolvers))
+	for i, resolver := range resolvers {
+		m, err := matcher.NewResolverMatcher(resolver.ComponentName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create matcher for resolver %d: %w", i, err)
+		}
+		matchers[i] = m
+	}
+
+	return &ResolverRepository{
+		goRoutineLimit: options.GoRoutineLimit,
+
+		repositoryProvider: repositoryProvider,
+		credentialProvider: credentialProvider,
+
+		resolvers:                    resolvers,
+		repositoriesForResolverCache: make([]repository.ComponentVersionRepository, len(resolvers)),
+		matcherCache:                 matchers,
+	}, nil
+}
+
+func (r *ResolverRepository) GetComponentVersion(ctx context.Context, component, version string) (*descriptor.Descriptor, error) {
+	repos := r.RepositoriesForComponentIterator(ctx, component)
```

Comment:
why do we need to iterate through repositories if this resolver can be deterministic at load time?

Reply:
I think it would be much simpler to load the order once, not cache and then use that order statically. I dont see a lot of value in lazy loading the repos. just my opinion though, happy to hear your thoughts